src/common/state_machine.hpp

```diff
@@ -223,6 +223,21 @@ class StateMachineRunner : virtual public EventPoster<EventType> {
 			// actionable until at least one state machine reaches a different state.
 			PostToEventLoop();
 		}
+
+		// Check if there are any non-deferred events in the queue - then fail if
+		queue<EventType> queue_copy = event_queue_;
+		while (not queue_copy.empty()) {
```

Comment:
It's expected to have non-deferred events here, because this is after the handlers have run. I think the check needs to be between the `if (!to_run.empty()) {` block and the one above it.

There is also a much easier way to check this. When we are doing `event_queue_.push(event)` above, also increment a `deferred_count` counter. If `deferred_count < event_queue_.size()`, then the check is positive.

Reply:
Oh! Very good idea!

I had a suspicion you would come up with something smarter than me here :laughing:

Reply:
Alright, new try. I am still using my original approach, but moved it to before we run the machine loop.

Ur approach, although clever, did not work when initializing (since we add the start events to the to_run queue, and I didn't feel like special casing it.

Reply:
This can be easily fixed too. Just keep a counter for the `run_queue` as well, but don't increment it during the initialization step, only when adding events in the "normal" flow. Then the check becomes `deferred_count < added_to_run_queue`. I think it's better than copying the whole queue on every state transition.

Reply:
True, I did dabble with the idea. But decided it's not really better. This implementation reads pretty straight forward, and does not have to touch the other pieces of code, and handle special cases.